### ISSUE TITLE
MVE

### ISSUE THREAD

**Issue by fabiencastan** [openMVG/openMVG#325]
Is there a way to use MVE with the new version?

**Comment by fabiencastan** [comment]
Link to the import code:
https://github.com/simonfuhrmann/mve/blob/master/apps/makescene/makescene.cc#L359

Check the "views.txt" here:
https://github.com/simonfuhrmann/mve/blob/master/apps/makescene/makescene.cc#L574

**Comment by pmoulon** [comment]
Here some information about how to write the openMVG2MVE exporter: https://github.com/simonfuhrmann/mve/issues/148#issuecomment-114061043

**Comment by mojovski** [comment]
I have already done some kind of MVE exporter but included the MVE libraries for this.
I guess this is not intended when an exporter from OpenMVG to MVE is in mind?

**Comment by pmoulon** [comment]
Hi @mojovski, as mentionned here by @simonfuhrmann https://github.com/simonfuhrmann/mve/issues/148#issuecomment-118255617 the exporter for MVE should be in OpenMVG.
We must add a openMVG2MVE binary that export an ASCII compatible MVE scene.
Does your exporter is working for the new format or for the old one?

There is no problem to depends on some optionnal library, we can check with CMAKE if the library is present or not and compile the openMVG2MVE only if MVE library is detected.

Normally @KevinCain plan to take a look to the exporter too...perhaps we could us three collaborate on it.

**Comment by KevinCain** [comment]
Hello @mojovski, it's good to hear you've already been active on this topic.  Here is the approach I began to sketch last week:

**Proposed openMVG2MVE2 (openMVG0.8-->MVE2) call**

*# Convert openMVG0.8 output to MVE2 file format
./ openMVG_main_openMVG2MVE2 -i Dataset/outGlobalSfM/← SfM_output/ -o Dataset/outGlobalSfM/SfM_output/*

**Proposed Implementation**

* seek openMVG root image path, dimensions, list of file names, IDs, camera intrinsic indices from user input path’s ‘sfm_data.json’ file (data version 0.1).
* iterate through ‘id_view’ list:
  * extract view intrinsics index, lookup value (see note below)
  * extract view file name, compute 50x50 thumbnail PNG via ImageMagick or similar
  * load views, poses, intrinsics, landmarks into OpenMVG SfM data container via openMVG ‘struct SfM_Data’ (http://openmv*g.readthedocs.org/en/latest/openMVG/sfm/sfm/#a-generic-sfm-data-container)
  * push back new MVE view via MVE libs -- _**what is the correct way to do this?**_
* save all views in native MVE2 format via MVE ‘void rewrite_all_views’:
 - directory "/views" contains the views that comprise the MVE scene
 - file ‘synth_n.out’ (where ‘n’ is the bundle #) bundle file contains key points

**Questions/Discussion**

On initialization with ‘makescene’ and processing with ‘sfmrecon’, MVE2 view directories contain:
- meta.ini:  camera intrinsics etc. (see sample below)
- image set:  original.jpg, thumbnail.png, **undistorted.png**
  - the file in bold above is created by ‘sfmrecon’; I can’t find a similar radial distortion corrected image written to disk by OpenMVG.  Am I missing something obvious?
 - exif.blob:  binary EXIF encoding?
  - how should we transfer/encode this information?
- (depth-L1.mvei):  Should we write this MVE image from OpenMVG or is it generated by dmrecon for depth illustration purposes?  I believe we can ignore this.

**Camera intrinsics**

I believe the best match for the MVE2 camera format, among OpenMVG’s pinhole three camera models, is ‘Pinhole_Radial_K1’, while the default in native OpenMVG scene’s appears to be ‘Pinhole_Radial_K3’.  The former is defined by image size (XY), focal length (), principal point (YX) and single-term radial distortion factor (as in MVE2).  We should leave a future path to carry 3-term distortion from OpenMVG in the event of a change to radial distortion correction in MVE2 at some future point.  In MVE native scenes generated via ‘makescene’, _the principal_point variable logged in ‘meta.ini’ is invariably 0.5 0.5 (center of image)._

**Note**:  It's worth pointing out that the MVE file format link given above in Simon's thread (https://github.com/simonfuhrmann/mve/wiki/The-MVE-File-Format) should be:  https://github.com/simonfuhrmann/mve/wiki/MVE-File-Format

**Sample 'meta.ini'**

#MVE view meta data is stored in INI-file syntax.
#This file is generated, formatting will get lost.

[camera]
focal_length = 0.8444210887
pixel_aspect = 1
principal_point = 0.5 0.5
rotation = 0.962009728 -0.1120776683 -0.2489494234 0.1936523616 0.9228841662 0.3328417838 0.1924473345 -0.3684066534 0.9095275998
translation = 2.006899118 -3.936580658 3.825217962

[view]
id = 0
name = IMG_0445

**Comment by simonfuhrmann** [comment]
I have a few suggestions. First, I would not bother using the MVE API to create the views. Because the new view format is so easy, there should be no need. Just generate the `meta.ini` file using standard file output. The `original.jpg` is optional and not required because OpenMVG will provide an undistorted image (ask @pmoulon how to get it). 

Also, it does not matter which camera model (or, distortion model) is used by OpenMVG, because the radial distortion parameters are not important if an undistorted image is already provided to MVE. Use the model that works best in OpenMVG, as long as it's pinhole. I'm sure OpenMVG has functions to create the thumbnail, i.e., a combination of cropping and rescaling.

Finally, the bundle file needs to be created, which might be the most tricky part, because it is important to get the features and the references to the images right. Otherwise MVS will not work.

Feel free to ask questions if anything is unclear.

**Comment by pmoulon** [comment]
Hi all,
Happy to see tha the conversation is so active ;-) and with clear question and statements

I agree with Simon, the OpenMVG2MVE exporter should export undistorted images. So, whatever the camere model used in OpenMVG a classic pinhole camera with no distortion will be provided to MVE.
For the principal point you should just use the openMVG camera pinhole divided by the image respective length.

In order to export undistorted image you can take a look to this binary https://github.com/openMVG/openMVG/blob/develop/src/software/SfM/main_ExportUndistortedImages.cpp

As the new format is ASCII only, normally MVE API must not be required.

In order to create thumbnail, there is new function in the develop branch about image sampling.
https://github.com/openMVG/openMVG/blob/develop/src/openMVG/image/image_resampling_test.cpp#L24
Basically you create a sampling grid and with that you can create tiny patches.
Or you can halfsample the image recusively:
https://github.com/openMVG/openMVG/blob/develop/src/openMVG/image/image_resampling.hpp#L20

I'm in holidays this week and unable to help. But next week I will be able to help more.

**Comment by KevinCain** [comment]
I've found the time to write a first draft of the converter, which I've added to the makefile on my openMVP devel branch to build with the rest of the software.  I call the new `openMVG_main_openMVG2MVE2` binary as follows:

`./openMVG_main_openMVG2MVE2 -i ../reconstruction_global/sfm_data.json -o /output_path/`

The exporter currently does the following:
 - Creates an appropriately named MVE2 directory for each view
 - Undistorts and saves each source image in the correct place
 - Writes 'meta.ini' with translated values (focal length, rotation matrix, translation vector, et cetera)
 - Writes the required camera information to the `synth_0.out` bundle

The next steps:
 - Write feature data to the `synth_0.out` bundle:  indexed view IDs and Feature IDs
 - Test value conversions with real-world cases

A few questions:

*I see the points stored in `sft_data.json' but don't know how to access them.*  Currently I'm using the `feats_provider` interface in openMVG to retrieve features, but it seems that is only used for storing U, V pixel addresses for rather than 3D points.  Also, I don't see how to access RGB values for these points in openMVG, which are included in the MVE2 output.

Once I understand the above well enough to finish a draft I'll send a pull request in this thread.

**Comment by pmoulon** [comment]
Hi @KevinCain 
Nice to hear that you have made progress.

Here a brief answer to your questions:
In order to access the 3D points you have to parse the sfm_data structure like here:
https://github.com/openMVG/openMVG/blob/develop/src/openMVG/sfm/sfm_data_io_ply.hpp#L61
Just get the structure field and loop over the 3D points and their 2D observations once you have read the SfM_Data scene.

```
SfM_Data sfm_data;
if (!Load(sfm_data, sSfM_Data_Filename, ESfM_Data(ALL))) {
  std::cerr << std::endl
    << "The input SfM_Data file \""<< sSfM_Data_Filename << "\" cannot be read." << std::endl;
  return EXIT_FAILURE;
}
const Landmarks & landmarks = data.GetLandmarks();
for (Landmarks::const_iterator iterLandmarks = landmarks.begin(); iterLandmarks != landmarks.end(); ++iterLandmarks)  {
  const Vec3 pointX = iterLandmarks->second.X;
}
```

- RGB values:
 - does not really matter, just put a white point for the moment (255 255 255).
- Radial distortion:
 - dist factor must be set to 0, since we export image with removed distortion

Hoping I provided answers to all your points.

**Comment by KevinCain** [comment]
Thank you @pmoulon, your answers are quite clear.  At some point if MVE's format evolves to handle polynomial distortion parameters, I do think it would be helpful to capture openMPV's radial distortion coefficients to MVE can recompute distortion from the source images rather than rely on the undistorted images -- but for now, without that functionality, you're right to set the values to 0 to reflect the corrected images.

I'll use your `sfm_data` structure as you suggest to iterate through the full set of 3D points in the bundle and a list of the 2D feature references by view as MVE expects.  Here's the relevant snippet from MVE's bundle export code for reference:

>   // For each feature, write to bundle:  position XYZ[0-3], color RGB[0-2], all ref.view_id & ref.feature_id
>    for (std::size_t i = 0; i < features.size(); ++i)
>    {
>        Bundle::Feature3D const& p = features[i];
>        out << p.pos[0] << " " << p.pos[1] << " " << p.pos[2] << "\n";
>        out << static_cast<int>(p.color[0] * 255.0f + 0.5f) << " "
>            << static_cast<int>(p.color[1] * 255.0f + 0.5f) << " "
>            << static_cast<int>(p.color[2] * 255.0f + 0.5f) << "\n";
>        out << p.refs.size();
>        for (std::size_t j = 0; j < p.refs.size(); ++j)
>        {
>            Bundle::Feature2D const& ref = p.refs[j];
>            out << " " << ref.view_id << " " << ref.feature_id << " 0";
>        }
>        out << "\n";
>    }

**Comment by pmoulon** [comment]
Here some additional code snippet that could help you:
```
const Landmarks & landmarks = data.GetLandmarks();
for (Landmarks::const_iterator iterLandmarks = landmarks.begin(); iterLandmarks != landmarks.end(); ++iterLandmarks)  {
  const Vec3 pointX = iterLandmarks->second.X;
  const Observations & obs = iterTracks->second.obs;
  for (Observations::const_iterator itObs = obs.begin(); itObs != obs.end(); ++itObs)
  {
    const IndexT viewId = itObs.first;
    const IndexT featId = itObs.second.id_feat;
  }
}
```

**Comment by simonfuhrmann** [comment]
Don't bother with the distortion coefficients. MVE will not do anything with it. I agree that keeping this information might prove useful potentially, you would also have to store information about which distortion model has been used, and formulas, and so on. 

Imported is that undistorted images are provided.

**Comment by KevinCain** [comment]
I've checked in my draft conversion code -- please see the pull request above.

As I wrote in the pull request, the MVE2 file output formatting seems fine as `dmrecon` is able to recognize and load the proper number of cameras and features, but `DMRecon::globalViewSelection()` finds no neighbors, reporting ‘Global View Selection failed’.

I suspect I am making a fundamental, though perhaps simple, error in encoding.  The code is very straightforward and uses, as much as possible, the standard methods I've gleaned from the openMPV source.

Here's an image of sample run of the converter using openMPV's inbuilt test data for Chateau de Sceaux:
![image](https://cloud.githubusercontent.com/assets/5424003/9080023/cad87804-3b04-11e5-826c-8c5014a9e0e3.png)

**Comment by KevinCain** [comment]
Thanks for the timely feedback, I'm happy to say that the MVE+FSSR output is looking much better now -- here is the OpenMVG sample data (Chateau de Sceaux, 11 photos) processed at level 1 sampling:
![output](https://cloud.githubusercontent.com/assets/5424003/9106519/b0e70b96-3bd5-11e5-8837-7ca35231f4aa.jpg)
If you have time to test with your own data that would make a nice next step.

**Comment by KevinCain** [comment]
For the moment I've hard coded a pixel aspect ratio of unity, but what it the correct way to poll OpenMVG for this value?

**Comment by KevinCain** [comment]
My latest commits are visible under the pull request above:
https://github.com/openMVG/openMVG/compare/develop...KevinCain:develop?expand=1

**Comment by simonfuhrmann** [comment]
I am pretty sure OpenMVG is not optimizing the pixel aspect and assumes 1. It is reasonable to hard code 1, or better, don't specify it and rely on MVEs default, which is of course also 1.

**Comment by pmoulon** [comment]
I will check the exporter on some scene during the day.
Will keep you informed ;-)

**Comment by pmoulon** [comment]
Thanks to @KevinCain and @simonfuhrmann the exporter seems to be almost ready. 
I will accept the pull request during the afternoon and work later on thumbnail generation (the only point that will be "hard" will be the aspect ratio conservation during the thumbnail computation)

![snapshot00](https://cloud.githubusercontent.com/assets/3235050/9133677/d54d7a00-3cfd-11e5-984c-05e4199f1503.jpg)

Ran on https://github.com/rperrot/ReconstructionDataSet/tree/master/MirebeauStHilaireStatue
![snapshot00](https://cloud.githubusercontent.com/assets/3235050/9133678/d934ef0e-3cfd-11e5-8fc7-d17e45c0b7e9.jpg)

![snapshot00](https://cloud.githubusercontent.com/assets/3235050/9133691/f71aec3a-3cfd-11e5-8477-8cb444688faa.jpg)


@fabiencastan You will be happy to see that the exported is functional ;-)

**Comment by fabiencastan** [comment]
Great! I will to test that ASAP.
Thanks to all the contributors!

**Comment by pmoulon** [comment]
@fabiencastan It's ready !
Just checkout latest openMVG develop branch and MVE /mve/tree/mve2 branch and have fun ;-)
See here for command line instructions:
https://github.com/openMVG/openMVG/blob/develop/docs/sphinx/rst/software/SfM/MVS.rst#export-to-mve-multi-view-environment

I close the issue since it is linked to  #351 and #355
